From c5380a1efc02d90d2b2384d003766d332264c7a2 Mon Sep 17 00:00:00 2001
From: Harini Katakam <harini.katakam@xilinx.com>
Date: Fri, 31 Jan 2020 21:07:18 +0530
Subject: [PATCH 1095/1851] net: macb: Handle phy reset for PCS PMA IP

commit cdedcb1df4d4928dfc89ca92891e0f414a96c5e9 from
https://github.com/Xilinx/linux-xlnx.git

Call phy reset (conditional on the flag the phy framework checks)
in macb_open and if the reset succeeds, re-initialize the phy.
This is required because of runtime clock handling.

Signed-off-by: Harini Katakam <harini.katakam@xilinx.com>
Reviewed-by: Radhey Shyam Pandey <radhey.shyam.pandey@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/net/ethernet/cadence/macb_main.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/ethernet/cadence/macb_main.c b/drivers/net/ethernet/cadence/macb_main.c
index 47bf50d81a1d..7b4e0ad4b5a8 100644
--- a/drivers/net/ethernet/cadence/macb_main.c
+++ b/drivers/net/ethernet/cadence/macb_main.c
@@ -2553,6 +2553,12 @@ static int macb_open(struct net_device *dev)
 	bp->macbgem_ops.mog_init_rings(bp);
 	macb_init_hw(bp);
 
+	/* Since this driver uses runtime handling of clocks, initiate a phy
+	 * reset if the attached phy requires it. Check return to see if phy
+	 * was reset and then do a phy initialization.
+	 */
+	if (phy_reset_after_clk_enable(dev->phydev) == 1)
+		phy_init_hw(dev->phydev);
 	/* schedule a link state check */
 	phy_start(dev->phydev);
 
-- 
2.31.1

